### ISSUE TITLE
Running into RangeOverflow Error when setting up the liana

### ISSUE THREAD

**Issue by roopakv** [ForestAdmin/forest-express-mongoose#113]
## Expected behavior

I expect the liana to not throw errors and add the `/forest` route to my app.

## Actual Behavior

It throws an error.

```
[forest] 🌳🌳🌳  An error occured while computing the Forest apimap. Your application apimap cannot be sent to Forest. Your Admin UI might not reflect your application models.
RangeError: Maximum call stack size exceeded
    at /Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:70:39
    at Array.forEach (<anonymous>)
    at unflatten (/Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:70:23)
    at /Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:98:23
    at Array.forEach (<anonymous>)
    at unflatten (/Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:70:23)
    at /Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:98:23
    at Array.forEach (<anonymous>)
    at unflatten (/Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:70:23)
    at /Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:98:23
    at Array.forEach (<anonymous>)
    at unflatten (/Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:70:23)
    at /Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:98:23
    at Array.forEach (<anonymous>)
    at unflatten (/Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:70:23)
    at /Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:98:23
    at Array.forEach (<anonymous>)
    at unflatten (/Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:70:23)
    at /Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:98:23
    at Array.forEach (<anonymous>)
    at unflatten (/Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:70:23)
    at /Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:98:23
    at Array.forEach (<anonymous>)
    at unflatten (/Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:70:23)
    at /Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:98:23
    at Array.forEach (<anonymous>)
    at unflatten (/Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:70:23)
    at /Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:98:23
    at Array.forEach (<anonymous>)
    at unflatten (/Users/roopak/workspace/repo/node_modules/forest-express-mongoose/node_modules/flat/index.js:70:23)
From previous event:
    at Object.perform (/Users/roopak/workspace/repo/node_modules/forest-express/dist/generators/schemas.js:14:14)
    at /Users/roopak/workspace/repo/node_modules/forest-express/dist/index.js:152:20
    at __NR_wrappedThenHandler (/Users/roopak/workspace/repo/node_modules/newrelic/lib/instrumentation/promise.js:468:25)
    at runCallback (timers.js:789:20)
    at tryOnImmediate (timers.js:751:5)
    at processImmediate [as _immediateCallback] (timers.js:722:5)
From previous event:
    at Promise.__NR_wrappedThen [as then] (/Users/roopak/workspace/repo/node_modules/newrelic/lib/instrumentation/promise.js:426:23)
    at Object.exports.init (/Users/roopak/workspace/repo/node_modules/forest-express/dist/index.js:149:50)
    at Object.exports.init (/Users/roopak/workspace/repo/node_modules/forest-express-mongoose/index.js:167:20)
    at Object.<anonymous> (/Users/roopak/workspace/repo/server/web.js:30:44)
    at Module._compile (module.js:635:30)
    at Object.Module._extensions..js (module.js:646:10)
    at Module.load (module.js:554:32)
    at tryModuleLoad (module.js:497:12)
    at Function.Module._load (module.js:489:3)
    at Function.wrappedLoad [as _load] (/Users/roopak/workspace/repo/node_modules/newrelic/lib/shimmer.js:342:38)
    at Module.require (module.js:579:17)
    at require (internal/module.js:11:18)
    at Object.<anonymous> (/Users/roopak/workspace/repo/index.js:25:1)
    at Module._compile (module.js:635:30)
    at Object.Module._extensions..js (module.js:646:10)
    at Module.load (module.js:554:32)
    at tryModuleLoad (module.js:497:12)
    at Function.Module._load (module.js:489:3)
    at Function.Module.runMain (module.js:676:10)
    at startup (bootstrap_node.js:187:16)
    at bootstrap_node.js:608:3
```

*Please provide any relevant information about your setup.*

* Package Version: ^2.13.1
* Express Version: ^4.15.4
* Mongoose Version: ^4.11.11
* MongoDB Version: 3.4

**Comment by arnaudbesnier** [comment]
Hi @roopakv, this is definitely an unexpected error.
In order to have more context and help you us identify what could happen can you provide a few more details:
- Did you follow the setup steps as described in the documentation/onboarding pages?
- Did you do some more actions?
- Do you have a lot of Mongoose schemas or specific schemas with a very large/complex/deep schema?

I imagine that this exception is due to some specific schemas but we need to identify them to go deeper.
We have several way to investigate on this:
- You can try to reduce the number of schema declared in your `modelsDir` to isolate the model(s) that raise the error.
- Even simpler you can use an undocumented property `includedModels` that makes the agent do the introspection only on the declared models. Something like:
```javascript
app.use(require('forest-express-mongoose').init({
  modelsDir: __dirname + '/models',
  envSecret: process.env.FOREST_ENV_SECRET,
  authSecret: process.env.FOREST_AUTH_SECRET,
  mongoose: require('mongoose') // The database connection.
  includedModels: ['customers', 'purchases', ...] // To replace with your own schema names
}));
```
should work.

**Comment by roopakv** [comment]
Sorry i was out for a couple of days.

- Did you follow the setup steps as described in the documentation/onboarding pages?
  - Yes
- Did you do some more actions?
  - No
- Do you have a lot of Mongoose schemas or specific schemas with a very large/complex/deep schema?
  - Big yes

Let me look into finding the model with a problem and get back to you folks. We are working to hit a deadline so expect me to get back to you toward the end of the week.

Also is there anyway we can use lumber with mongodb? or any solution where we keep our data instead of pushing it to your servers? basically a hosted solution?

**Comment by roopakv** [comment]
So I did some work around this.

-> even if I use includedModels since all models are already loaded into mongoose, the above error is thrown. It seems to be a problem in `forest-express`.

Is there an easy way to load model by model rather than with `forest-express-mongoose `

**Comment by arnaudbesnier** [comment]
Hi @roopakv, As I said in my previous comment, the other way to investigate is to remove all your schema declaration files and re-add them progressively to isolate the schemas that generate such error.

**Comment by saintmac** [comment]
@arnaudbesnier I've got the same error, with a dead simple setup: only 1 model with 1 Boolean field (also tested with 1 String field and got the same error).
I tested with a few recent versions of express, mongoose and this package.
Including: `mongoose@5.3.0`, `express@4.16.3` and `forest-express-mongoose@2.14.1`

**Comment by bguivarch** [comment]
@arnaudbesnier I have the same error as well.
`forest-express-mongoose@2.14.1`, `mongoose@5.2.7` an `express@4.15.3`

**Comment by ArnaudValensi** [comment]
Hi everybody, I'm on it.

**Comment by ArnaudValensi** [comment]
Hi @roopakv, @saintmac, @bguivarch,
The issue is fixed in the version `2.14.2`.
You can do:
```
npm install forest-express-mongoose@latest
```
or
```
yarn add forest-express-mongoose@latest
```

**Comment by roopakv** [comment]
Thank you so much Arnaud! ghat worked

**Comment by saintmac** [comment]
Working for me as well, thanks!

**Comment by ArnaudValensi** [comment]
Sorry guys, we had to release a version `2.14.4` of `forest-express-mongoose` with the patch for this issue reverted. It was causing some troubles with the embedded documents.

So the recommendations for now are the following:
- Use the version `2.14.4` with `mongoose` `5.2.*`
- Use the version `2.14.2` or `2.14.3` if you really need to use the latest version of mongoose, but you may have troubles with subdocuments

I'm currently working on a definitive fix. I'll keep you posted.

**Comment by roopakv** [comment]
Thank you for the update. it seems like 2.14.4 with mongoose 5.2.14 doesnt seem to work for us.

Looking forward to see what fix you have planned

**Comment by ArnaudValensi** [comment]
This is now fixed in the `2.14.5`. It also solve a problem where sometime internal mongoose fields are considered as regular collection fields.
